### DISCUSSION TITLE
u-boot in Jetson AGX Xavier

### DISCUSSION THREAD

**Discussion by armhzjz** [OE4T/meta-tegra#564] [Q&A] [answered]
Hello,
I built dunfell-l4t-r32.4.3. Initially I wanted to prepare my image so that u-boot can fetchthe image from tftp (grabbing the Image from an external volume) and then preparing the kernel so that it uses an exteral rootfs via nfs.
With this in mind I tried to explore the u-boot recipes but I didnt find them on my build. I look for them using `devtool search u-boot` and then `devtool search tegra-boot`, but see no recipe for the u-boo bootloader.
Now I am wondering if the Jetson AGX Xavier image includes (or can include) u-boot.
Is that possible? If so, do I have to explicitly set some configuration in my local.conf to enable/include u-boot?

**Comment by ichergui** [comment, marked as answer]
Hello @armhzjz 
I don't that this is possible with the Jetson AGX Xavier because there is no support for `u-boot`.
Please check the Boot flow as Nvidia describe [here](https://docs.nvidia.com/jetson/archives/l4t-archived/l4t-325/index.html#page/Tegra%20Linux%20Driver%20Package%20Development%20Guide/bootflow_jetson_xavier.html)
Tbh, I didn't try this.
I think, this is doable by using `cboot`. 
Let's see if @madisongh have some thoughts about it.

**Comment by madisongh** [comment]
That's correct, there is no U-Boot for the Xavier platforms.

cboot is supposed to support TFTP booting, although I've never tried it.  Check the [Boot Options](https://docs.nvidia.com/jetson/l4t/index.html#page/Tegra%20Linux%20Driver%20Package%20Development%20Guide/getting_started.html#wwpID0E0XI0HA) page in the L4T docs for information on setting it up.

**Comment by armhzjz** [comment]
Hello @ichergui and @madisongh,
Thanks for the hints. Already looking into  it.

**Comment by coderbusy7** [comment]
@armhzjz  how did you manage to use cboot for xavier to build an image? 
 Have you been able to inherit certain image from the poky reference images? I've tried to do the same but build is failing because of u-boot. I'm hitting the following error
`  "error: could not load ramdisk"`

**Comment by ichergui** [comment]
Hey @anishmonachan7 

`Jetson Xavier` use by default `cboot`. Are you using [tegra-demo-distro](https://github.com/OE4T/tegra-demo-distro) ? if, yes
You don't have to do anything.

Please check the boot flow of `Jetson Xavier` [here](https://docs.nvidia.com/jetson/archives/l4t-archived/l4t-325/index.html#page/Tegra%20Linux%20Driver%20Package%20Development%20Guide/bootflow_jetson_xavier.html)

**Reply by coderbusy7** [reply]
Thanks for the responce @ichergui. I'm not using the tegra-demo-distro for now. Do you know if it's compatible with warrior? i didn't see a warrior branch

**Reply by ichergui** [reply]
[tegra-demo-distro ](https://github.com/OE4T/tegra-demo-distro) is not compatible with `warrior` Yocto release. Our stable branch is `dunfell-l4t-r32.4.3`, you can take a look [here](https://github.com/OE4T/meta-tegra/wiki/Which-branch-should-I-use%3F) concerning the branch you should use.

**Reply by coderbusy7** [reply]
@ichergui I'm wondering if we have any other solution rather switching to dunfell branch?

**Comment by armhzjz** [comment]
Hello @anishmonachan7 
I didn't need to do anything. If you try to build meta-tegra and provided you are attempting to build for Xavier, the only thing you actually need to do is to set your machine variable in the local.conf file to the appropriate device (i.e. jetson-xavier if building branch *l4t-r32.4.* or jetson-agx-xavier-devkit if building branch *l4t-r32.5.0:
For example in your build/conf/local.conf:
`machine = "jetson-xavier"`

If building for Jeton Xavier CBoot will be built by default. And as far as I understand, u-boot is no supported at all on this device (i.e. on Xavier SoC).

**Comment by armhzjz** [comment]
@anishmonachan7  reading your quesiton for a second time I realize that maybe you were able to successfully build CBoot and that the error you mentinoed comes only after you boot your device. Is this assumption right?
If so, then unfortunatelly I have not a clue of why you could be getting this. Did you modify anything on the CBoot device tree?

**Reply by coderbusy7** [reply]
I haven't modified the image tree. I'm building the image using warrior branch. My build is failing with u-boot issues. Then I have used the following command to skip u-boot build. 
`bitbake -I u-boot core-image`

My end use case is to integrate mender with meta-tegra and mender needs u-boot support.

**Reply by ichergui** [reply]
Could you please share the build issue with `u-boot`  ? 
Mender is working fine with `Jetson's Xavier`

**Reply by madisongh** [reply]
There is no u-boot for Xavier platforms. If you look at the `meta-mender-tegra` layer in the [meta-mender-community](https://github.com/mendersoftware/meta-mender-community) repo, you'll find the combination of U-Boot utils emulation and state scripts that were used to provide integration for cboot-based platforms (like the [Xavier NX](https://hub.mender.io/t/nvidia-tegra-jetson-xavier-nx/2615) and [AGX Xavier](https://hub.mender.io/t/nvidia-tegra-agx-xavier/2616)). Support was added for dunfell and later, but you may be able to back-port it to warrior if you're really stuck on that obsolete version.

**Reply by coderbusy7** [reply]
Thank you @madisongh @armhzjz @ichergui for your quick responses and help.
